main.cpp

```diff
@@ -279,387 +224,205 @@ struct HeapA
+        *value = *value / other;
+        return *this;
     }
-    return *this;
-}
-
-IntType& IntType::powInternal(int rhs)
-{
-    *value = static_cast<int>(std::pow( *value, rhs ));
-    return *this;
-}
 
-IntType& IntType::pow(const FloatType& ft)
-{
-    return powInternal(static_cast<int>(static_cast<float>(ft)));
-}
-
-IntType& IntType::pow(const IntType& it)
-{
-    return powInternal(static_cast<int>(it));
-}
-
-IntType& IntType::pow(const DoubleType& dt)
-{
-    return powInternal(static_cast<int>(static_cast<double>(dt)));
-}
+    Numeric& pow(const Type& dt)
+    {
+        return powInternal(static_cast<double>(dt));
+    }
 
-IntType& IntType::pow(int i_)
-{
-    return powInternal(i_);
-}
+    template<typename Callable>
+    Numeric& apply(Callable&& doubleFptr)
+    {
+        doubleFptr(value);
+        return *this;
+    }
 
-DoubleType& DoubleType::powInternal(double rhs)
-{
-    *value = std::pow( *value, rhs );
-    return *this;
-}
+private:
+    std::unique_ptr<Type> value;
 
-DoubleType& DoubleType::pow(const FloatType& ft)
-{
-    return powInternal(static_cast<double>(static_cast<float>(ft)));
-}
+    Numeric& powInternal(Type rhs)
+    {
+        *value = std::pow(*value, rhs);
+        return *this;
+    }
+};
 
-DoubleType& DoubleType::pow(const IntType& it)
+template<typename NumericType>
+void add7(std::unique_ptr<NumericType>& val)
 {
-    return powInternal(static_cast<double>(static_cast<int>(it)));
+    {
+        *val += static_cast<NumericType>(7.0f);
+    }
 }
 
-DoubleType& DoubleType::pow(const DoubleType& dt)
+template<typename NumericType>
+void add5(std::unique_ptr<NumericType>& val)
 {
-    return powInternal(static_cast<double>(dt));
+    {
+        *val += 5;
+    }
 }
 
-DoubleType& DoubleType::pow(double d_)
+template<typename NumericType>
+void add6(std::unique_ptr<NumericType>& val)
 {
-    return powInternal(static_cast<double>(d_));
+    {
+        *val += 6;
+    }
 }
 
 struct Point
 {
-    Point(const float& fx_, const float& fy_) : x(fx_), y(fy_) {}
-
-    Point(const FloatType& ftx_, const FloatType& fty_) : Point(static_cast<float>(ftx_), static_cast<float>(fty_)) {}
-    Point(const IntType& itx_, const IntType& ity_) : Point(static_cast<float>(static_cast<int>(itx_)), static_cast<float>(static_cast<int>(ity_))) {}
-    Point(const DoubleType& dtx_, const DoubleType& dty_) : Point(static_cast<float>(static_cast<double>(dtx_)), static_cast<float>(static_cast<double>(dty_))) {}
+    template<typename Type1, typename Type2>
+    
+    Point(const Type1& tx_, const Type2& ty_) : x(tx_), y(ty_) {}
     
-    Point& multiply(float m);
-    Point& multiply(FloatType& ftm);
-    Point& multiply(DoubleType& dtm);
-    Point& multiply(IntType& itm);
+    Point& multiply(float m)
+    {
+        x *= m;
+        y *= m;
+        return *this;
+    }
 
-    void toString();
+    template<typename Type>
+    Point& multiply(Type& rhs)
```

Comment:
your function argument should be `const Type&`
Always pass by const-ref unless you need to modify the passed-in object.
Alternatively, pass by copy/value if the object is copyable. 
- [ ] revise